File: 000_utils/binaryTree/BinaryTree.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {};
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class BinaryTree {
    public TreeNode createTreeUsingBFS(int[] nums) {
        TreeNode[] nodeArr = new TreeNode[nums.length];
        for (int i = 0; i < nums.length; i++) {
            nodeArr[i] = new TreeNode(nums[i]);
            if (i == 0) continue;
            if (i % 2 == 1) nodeArr[(i - 1) / 2].left = nodeArr[i];
            else nodeArr[(i - 1) / 2].right = nodeArr[i];
        }

        return nodeArr[0];
    }

    public static void dfs(TreeNode root) {
        if (root == null) return;

        System.out.print(root.val + " ");
        dfs(root.left);
        dfs(root.right);
    }

    public static void bfs(TreeNode root) {
        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.addLast(root);
        while (!queue.isEmpty()) {
            TreeNode curNode = queue.pollFirst();
            System.out.print(curNode.val + " ");
            if (curNode.left != null) queue.addLast(curNode.left);
            if (curNode.right != null) queue.addLast(curNode.right);
        }
    }

    public static void main(String[] args) {
        int[] nums = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13};
        TreeNode root = new BinaryTree().createTreeUsingBFS(nums);
        bfs(root);
    }
}

File: 000_utils/binaryTree/BinaryTree.go
package main

import "fmt"

type TreeNode struct {
    Val   int
    Left  *TreeNode
    Right *TreeNode
}

func createTreeUsingBFS(nums []int) *TreeNode {
    arrLen := len(nums)
    var nodeArr []*TreeNode

    for i := 0; i < arrLen; i++ {
        nodeArr = append(nodeArr, &TreeNode{nums[i], nil, nil})
        if i == 0 { continue }
        if i % 2 == 1 {
            nodeArr[(i - 1) / 2].Left = nodeArr[i]
        } else {
            nodeArr[(i - 1) / 2].Right = nodeArr[i]
        }
    }

    return nodeArr[0]
}

func dfs(root *TreeNode) {
    if root == nil { return }
    fmt.Printf("%d ", root.Val)
    dfs(root.Left)
    dfs(root.Right)
}

func bfs(root *TreeNode) {
    queue := []*TreeNode{root}

    for len(queue) != 0 {
        curNode := queue[0]
        queue = queue[1:]
        fmt.Printf("%d ", curNode.Val)
        if curNode.Left != nil { queue = append(queue, curNode.Left) }
        if curNode.Right != nil { queue = append(queue, curNode.Right) }
    }
}


func main() {
    nums := []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13}
    root := createTreeUsingBFS(nums)
    bfs(root)
}
